### ISSUE TITLE
Weight matrix not updating with a user defined initial weight matrix

### ISSUE THREAD

**Issue by mk24934** [FluxML/Flux.jl#1717]
For reproducibility, I would like to run one epoch of gradient descent starting with a weight matrix of all ones. In the minimal working example below I call the `init` keyword argument in `Dense` with a user defined function `ones_function` that sets the initial weight matrix as ones. However after running `train!` on this model the weight matrix is not updating. 

Below is some code to reproduce this:

```julia
using Flux

x = Float64.([[1, 2, 3, 4] [4, 8, 10, 6]])
y = Float64.(reshape([3, 6, 8, 7], 4, 1)) # add a reshape in here to set this type as matrix

# Step 1: Make a one node neural network
# We want to set the initial weight matrix as all ones
# the init argument in `Dense` requires a function with two arguments
ones_function = function(x, y) ones(1, 2) end
linear_layer = Dense(2, 1; init = ones_function)
model_one = Chain(linear_layer)
print("Initial parameters are: ", params(model_one), "\n")

# Step 2: Define compilation items
loss_function(y_hat, y) = Flux.Losses.mse(y_hat, y)
parameters = params(model_one)
dataset = [(x, y)]
optimizer = Flux.Optimise.RMSProp(0.1)

# Step 3: train! for 1 epoch
Flux.@epochs 1 Flux.train!(
        loss_function,
        parameters,
        dataset,
        optimizer
)

# Step 4: Look at the updated weight matrix
print("After one epoch the parameters are: ", params(model_one))
```

**Comment by darsnack** [comment]
Your loss function doesn't incorporate your model which is why nothing is updating (the gradient with respect to your model's parameters is zero, because it is not being used).

Instead try
```julia
using Flux

x = Float32.([[1, 2, 3, 4] [4, 8, 10, 6]])
y = Float32.(reshape([3, 6, 8, 7], 4, 1)) # add a reshape in here to set this type as matrix

# Step 1: Make a one node neural network
# We want to set the initial weight matrix as all ones
# the init argument in `Dense` requires a function accepting dimensions of the array
ones_function(dims...) = ones(Float32, dims...)
linear_layer = Dense(2, 1; init = ones_function)
model_one = Chain(linear_layer)
print("Initial parameters are: ", params(model_one), "\n")

# Step 2: Define compilation items
loss_function(y_hat, y) = Flux.Losses.mse(y_hat, y)
parameters = params(model_one)
dataset = [(x, y)]
optimizer = Flux.Optimise.RMSProp(0.1)

# Step 3: train! for 1 epoch
Flux.@epochs 1 Flux.train!(
        (x, y) -> loss_function(model_one(x), y),
        parameters,
        dataset,
        optimizer
)

# Step 4: Look at the updated weight matrix
print("After one epoch the parameters are: ", params(model_one))
```

Note I also changed your `ones_function` slightly. The `init` keyword expects a function that takes the dimensions of the array as input. In the case of `Dense`, this will be two values (which is why your original code worked), but for arbitrary layers, it could be more. The new `ones_function` accepts any number of arguments.

**Comment by DhairyaLGandhi** [comment]
Closing this since it isn't a flux issue.
